### ISSUE TITLE
Mocha 7: Failing async test exits mocha watch mode on reload

### ISSUE THREAD

**Issue by mantoni** [mochajs/mocha#4143]
### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [x] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend that you _not_ install Mocha globally.

### Description

Mocha 7 exits when re-running a failing async test case.

### Steps to Reproduce

```js
describe('watch fails on reload', () => {

  it('fails', (done) => {
    setTimeout(() => {
      throw new Error();
    }, 1)
  });

});
```

Run `mocha --watch`, then touch the file to run the test again. Mocha 7 exists while mocha 6 continues to run.

**Expected behavior:**

Mocha shows the failure and continues to watch the file for changes.

**Actual behavior:**

Mocha 7 exits when re-running the test. Mocha 6 shows the expected bahavior with the above test case. Mocha 7 behaves as expected for synchronous test cases.

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

- Mocha v7.0.0
- Node v12.1.0
- Mac OS Catalina 10.15.2 (19C57)
- Fish shell

**Comment by juergba** [comment]
@mantoni thanks for reporting.

It's a badly written test by throwing an uncaught exception in an async scenario. Instead of calling `done(new Error())`.

Could you give the current `master` branch a try, please?

**Comment by HyunSangHan** [comment]
@juergba 
I tried the test with the above reproducing code in both versions(`v7.0.0` and the current `master` branch).
Mocha v7.0.0 exits when re-running the test as @mantoni mentioned, but the current master branch doesn't.

**Comment by sla100** [comment]
You should follow modern async/await syntax:
```js
import util from 'util';

it('fails', async () => {
   await util.promisify(setTimeout)(1);
   throw new Error();
}
```

**Comment by mantoni** [comment]
@juergba I usually don't write tests like this, I'm just demonstrating the issue. I'm doing TDD with Mocha watchmode and sometimes I have a typo that causes an exception.

**Comment by mantoni** [comment]
@sla100 Thank you for the suggestion. You're obviously allowed to write tests the way you want. I'm not suggesting that you should adobt my testing style. I'm just demonstrating an issue here.

**Comment by juergba** [comment]
@HyunSangHan thanks for testing.

fixed by #4147.

**Comment by mantoni** [comment]
Awesome! Thanks you 💯
